File: server/src/module/dmessage.js
import { calculatePP } from './performance';
import * as cheerio from 'cheerio';

export async function buildSolvedMessage(challengeId, solver, test = false) {
  let challengeInfo = null;

  try {
    const response = await fetch(`https://dreamhack.io/api/v1/wargame/challenges/${challengeId}/`);
    if (response.ok) {
      challengeInfo = await response.json();
    } else {
      throw Error(JSON.stringify(response));
    }
  } catch (error) {
    console.error('문제 정보 가져오기 실패:', error);
    return;
  }

  let valid = challengeInfo && solver;
  const rate = valid ? challengeInfo.cnt_solvers / challengeInfo.hitcount * 100 : 'N/A';
  const level_color = 1 <= challengeInfo.difficulty && challengeInfo.difficulty <= 3 ? 0x43B581 :
                      4 <= challengeInfo.difficulty && challengeInfo.difficulty <= 6 ? 0x09BAF9 :
                      7 <= challengeInfo.difficulty && challengeInfo.difficulty <= 8 ? 0x013CC7 :
                      9 === challengeInfo.difficulty ? 0xFC4749 :
                      10 === challengeInfo.difficulty ? 0xC90002 :
                      0xFFFFFF;

  const pp = calculatePP(challengeInfo);
  valid = pp ? valid : false;

  let nickname, introduction, profile_image;
  try {
    const response = await fetch(`https://dreamhack.io/users/${solver}`);
    if (!response.ok) {
      throw Error(JSON.stringify(response));
    }
    const $ = cheerio.load(await response.text(), 'text/html');
    const profile = $('.user-profile');
    nickname = profile.find('.nickname').text().trim().replace('`', "'");
    introduction = profile.find('.intro-text').text().trim().replace('`', "'");
    profile_image = profile.find('.user-icon > span > img').attr('src');
  } catch (error) {
    console.error(error);
    return;
  }

  if(!valid){
    console.error(solver, pp);
  }

  return {
    content: "",
    embeds: [{
      title: valid ? `🎉 ${challengeInfo.title} 문제 해결!` : `🎉 Challenge #${challengeId} 문제 해결!`,
      description: valid ?
        `**해결자**\n` + `[\`${nickname}\`](https://dreamhack.io/users/${solver})` +
        (introduction !== "아직 자기소개가 없습니다." ? ` | \`${introduction}\`\n\n` : '\n\n') +
        `**난이도**\n` + `LEVEL ${challengeInfo.difficulty}\n\n` +
        `**태그**\n` + `${challengeInfo.tags.map(tag => `#${tag}`).join(', ')}\n\n` +
        `**솔버 수**\n` + `${challengeInfo.cnt_solvers} solved / ${challengeInfo.hitcount} viewed` + (rate < 2 ? ` **(${rate.toFixed(2)}%)**\n\n` : ` (${rate.toFixed(2)}%)\n\n`) +
        `**예상되는 퍼포먼스**\n${pp.toFixed(2)}pp` + (test ? `\n\n*이 메시지는 웹훅 테스트 메시지이며, 실제로 풀이된 것이 아닙니다.*` : '')
        :
        `Challenge #${challengeId}를 해결했습니다!` + (test ? `\n\n*이 메시지는 웹훅 테스트 메시지이며, 실제로 풀이된 것이 아닙니다.*` : ''),
      color: level_color,
      url: `https://dreamhack.io/wargame/challenges/${challengeId}`,
      timestamp: new Date().toISOString(),

      // It is safe to use the raw solver input object because Dreamhack and Discord will block exploits.
      ...(solver ? { thumbnail: { url: profile_image || 'https://static.dreamhack.io/main/v2/img/amo.1a05d65.png' } } : {}),
    }]
  };
}

File: server/src/modules/dmessage.rs
use crate::modules::performance;
use crate::modules::request;
use crate::modules::types::*;
use anyhow::Result;
use dotenv::dotenv;
use indoc::indoc;
use once_cell::sync::Lazy;

static URL: Lazy<String> = Lazy::new(|| {
    dotenv().ok();
    std::env::var("DISCORD_WEBHOOK_URL").expect("DISCORD_WEBHOOK_URL must be set")
});

pub async fn build_solved_message(
    challenge_id: &u32,
    solver_id: &u32,
    test: &bool,
) -> Result<String> {
    let (challenge_info, solver) = tokio::join!(
        request::get_challenge(&challenge_id),
        request::get_user(&solver_id)
    );
    let challenge_info = challenge_info?;
    let solver = solver?;

    let rate = challenge_info.cnt_solvers as f64 / challenge_info.hitcount as f64 * 100.0;
    let level_color = match challenge_info.difficulty {
        1..=3 => 0x43B581,
        4..=6 => 0x09BAF9,
        7..=8 => 0x013CC7,
        9 => 0xFC4749,
        10 => 0xC90002,
        _ => 0xFFFFFF,
    };

    let pp = performance::calculate_pp(&challenge_info);

    let message = SolvedMessage {
        content: "".to_string(),
        embeds: vec![Embed {
            title: format!("🎉 {} 문제 해결!", challenge_info.title),
            description: format!(
                indoc! {
                    "**해결자**
                    [`{}`](https://dreamhack.io/users/{}){}

                    **난이도**
                    LEVEL {}

                    **태그**
                    {}

                    **솔버 수**
                    {} solved / {} viewed ({}{:.2}%{})

                    **예상되는 퍼포먼스**
                    {:.2}pp
                    {}
                    "
                },
                solver.nickname,
                solver_id,
                if solver.introduction != "아직 자기소개가 없습니다." {
                    format!(" | `{}`", solver.introduction)
                } else {
                    String::new()
                },
                challenge_info.difficulty,
                challenge_info.tags.join(", "),
                challenge_info.cnt_solvers,
                challenge_info.hitcount,
                if rate < 2.0 { "**" } else { "" },
                rate,
                if rate < 2.0 { "**" } else { "" },
                pp,
                if *test {
                    "\n\n*이 메시지는 웹훅 테스트 메시지이며, 실제로 풀이된 것이 아닙니다.*"
                } else {
                    ""
                }
            ),
            color: level_color,
            url: format!("https://dreamhack.io/wargame/challenges/{}", challenge_id),
            timestamp: chrono::Utc::now().to_rfc3339(),
            thumbnail: Image {
                url: solver.profile_image,
            },
        }],
    };

    Ok(serde_json::to_string(&message)?)
}

pub async fn send_message(message: String) -> Result<()> {
    let _ = reqwest::Client::new()
        .post(URL.to_owned())
        .header("Content-Type", "application/json")
        .body(message)
        .send()
        .await?;
    Ok(())
}
